keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md

```diff
@@ -0,0 +1,337 @@
+- Applications that process very important data. They should scale up as fast as possible (to reduce the data processing time), and scale down as soon as possible (to reduce the costs). False positives signals to scale up/down are ok.
+- Applications that process other data/web traffic. It is not that important, and may scale up and down in a usual way to minimize jitter.
+
+At the moment, there’s only one cluster-level configuration parameter that influence how fast the cluster is scaled down:
+
+- [--horizontal-pod-autoscaler-downscale-stabilization-window][]   (default to 5 min)
+
+And a couple of hard-coded constants that specify how fast the cluster can scale up:
+
+- [scaleUpLimitFactor][] = 2.0
+- [scaleUpLimitMinimum][] = 4.0
+
+As a result users can't influence the scale velocity and that is a problem for a lot of people. There're several open issues in tracker about that:
+
+- [#39090][]
+- [#65097][]
+- [#69428][]
+
+[--horizontal-pod-autoscaler-downscale-stabilization-window]: https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/#algorithm-details
+[scaleUpLimitFactor]: https://github.com/kubernetes/kubernetes/blob/v1.13.4/pkg/controller/podautoscaler/horizontal.go#L55
+[scaleUpLimitMinimum]: https://github.com/kubernetes/kubernetes/blob/v1.13.4/pkg/controller/podautoscaler/horizontal.go#L56
+[#39090]: https://github.com/kubernetes/kubernetes/issues/39090
+[#65097]: https://github.com/kubernetes/kubernetes/issues/65097
+[#69428]: https://github.com/kubernetes/kubernetes/issues/69428
+
+### Goals
+
+- Allow the user to be able to manage the scale velocity
+
+### Non-Goals
+
+TBA
+
+## Proposal
+
+We need to introduce an algorithm-agnostic HPA object configuration that will configure each particular HPA scaling behaviour.
+For both direction (scale up and scale down) there should be two parameters:
+
+- Parameter to specify the relative speed, in percentages:
+  - `ScaleUpPercent`
+    i.e. if ScaleUpPercent = 150 , then we can add 150% more pods (10 -> 25 pods)
+  - `ScaleDownPercent`
+    i.e. if ScaleDownPercent = 60 , then we can remove 60% of pods (10 -> 4)
+- Parameter to specify the absolute speed, in number of pods:
+  - `ScaleUpPods`
+    i.e. if ScaleUpPods = 5 , then we can add 5 more pods (10 -> 15)
+  - `ScaleDownPods`
+    i.e. if ScaleDownPods = 7 , then we can remove 7 pods (10 -> 3)
+
+All the parameters are per-minute and allow fraction values by using Quantity type.
+
+A user will specify the parameters for the HPA, thus controlling the HPA logic.
+
+If the user specify two parameters, two constraints are calculated, and the largest is used (see the [Motivation for pick the largest constraint][] section below).
+
+If the user doesn’t want to use some particular parameter (= user doesn’t want to use this particular constraint), the parameter should be set to -1 (or any negative number).
```

Comment:
Why?  Optional fields are explicitly not specified.  Sentinel values like this make for bad APIs.

Reply:
We don't have another option if we want to keep the API and make the previously created HPA work the same way.

Currently, you don't have a way to configure HPA scale velocity.
So for scaling up, it chooses the max of the following:
 - it either [doubles the number of pods](https://github.com/kubernetes/enhancements/pull/883/files/33127648cb3176b84992ea84e2b6f8df3b06ddcc#diff-fd841be6aa8e9bedba8f79659f684798R63)
 -  or it [adds 4 pods](https://github.com/kubernetes/enhancements/pull/883/files/33127648cb3176b84992ea84e2b6f8df3b06ddcc#diff-fd841be6aa8e9bedba8f79659f684798R64)

For scaling down, it works like there's no limitation in velocity, it picks the "desiredPods", and then applies the "Stabilization Window".
So it is equivalent to the following configuration (in terms of this KEP):
```
- constraints:
  - scaleUp:
    - percent = 100
    - pods = 4
  - scaleDown:
    - percent = -1
    - pods = -1
```

If we choose to use `-1` value as a default value, it will make all previously existed HPA (without `constraints` block) to work like it has the following configuration:
```
- constraints:
  - scaleUp:
    - percent = -1
    - pods = -1
  - scaleDown:
    - percent = -1
    - pods = -1
```

So, HPA will pick the "desiredPods" value instantly instead of limiting it in some way.

Reply:
AFAIR, we don't use such values anywhere in K8S and we should not make an exception here.

Reply:
Fixed.